### PR TITLE
Update for Android Gradle plugin  2.5

### PR DESCRIPTION
**Pull request by rschiu**
Use compileClasspath Configuration for each variant instead
of the compile Configuration for the source sets for the
extract include proto tasks.

### CODE REVIEW

**Comment by adrien-aubel** [comment]
Thanks @rschiu, this let me build with Android Studio 3.0 Canary 1 / Gradle 4.0-milestone-1 / build tools 3.0.0-alpha1.

Hopefully this will make it in a 0.8.2 released soon? :) (CC @zhangkun83)

**Comment by LouisCAD** [comment]
Why this PR still has not been merged almost a month later? My Android project is dying to try the new Android Studio!

**Comment by zhangkun83** [comment]
I am still waiting for @rschiu to post the latest changes to this PR.

**Comment by lucastsantos** [comment]
looks like it's rebased now 🙇

File: gradle/wrapper/gradle-wrapper.properties [hunk 1]
```diff
@@ -3,4 +3,4 @@ distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-3.0-all.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-4.0-rc-3-all.zip
```
Comment (zhangkun83) [inline gradle/wrapper/gradle-wrapper.properties]: Please add `4.0-rc-3` to `ProtobufJavaPluginTest.gradleVersions` and `ProtobufAndroidPluginTest`.

File: src/main/groovy/com/google/protobuf/gradle/ProtobufPlugin.groovy [hunk 2]
```diff
@@ -241,9 +244,29 @@ class ProtobufPlugin implements Plugin<Project> {
       sourceSetNames.each { sourceSetName ->
         def extractProtosTask = maybeAddExtractProtosTask(sourceSetName)
         generateProtoTask.dependsOn(extractProtosTask)
+      }
 
-        def extractIncludeProtosTask = maybeAddExtractIncludeProtosTask(sourceSetName)
+      // Extract proto file from the compile classpath configuration exposed in Android Gradle plugin 2.5.  Revert to
+      // original behavior if the configuration is not found.
+      Configuration compileConfiguration = variant.hasProperty("compileConfiguration") ? variant.compileConfiguration : null
+      if (compileConfiguration != null) {
```
Comment (zhangkun83) [inline src/main/groovy/com/google/protobuf/gradle/ProtobufPlugin.groovy]: `protobuf-gradle-plugin/testProjectAndroid/build.gradle` is still using `com.android.tools.build:gradle:2.2.0-beta1`. This new branch is not tested. You will need some work on the test code (`ProtobufAndroidPluginTest` and its friends) to test both the old and the new Android plugins.

File: src/main/groovy/com/google/protobuf/gradle/ProtobufPlugin.groovy [hunk 3]
```diff
@@ -241,9 +244,29 @@ class ProtobufPlugin implements Plugin<Project> {
       sourceSetNames.each { sourceSetName ->
         def extractProtosTask = maybeAddExtractProtosTask(sourceSetName)
         generateProtoTask.dependsOn(extractProtosTask)
+      }
 
-        def extractIncludeProtosTask = maybeAddExtractIncludeProtosTask(sourceSetName)
+      // Extract proto file from the compile classpath configuration exposed in Android Gradle plugin 2.5.  Revert to
+      // original behavior if the configuration is not found.
+      Configuration compileConfiguration = variant.hasProperty("compileConfiguration") ? variant.compileConfiguration : null
```
Comment (rschiu) [inline src/main/groovy/com/google/protobuf/gradle/ProtobufPlugin.groovy]: done

File: src/main/groovy/com/google/protobuf/gradle/ProtobufPlugin.groovy (hunk 2 above)
Comment (rschiu) [inline src/main/groovy/com/google/protobuf/gradle/ProtobufPlugin.groovy]: done

File: src/main/groovy/com/google/protobuf/gradle/ProtobufPlugin.groovy [hunk 4]
```diff
@@ -241,9 +244,29 @@ class ProtobufPlugin implements Plugin<Project> {
...
+      if (compileConfiguration != null) {
+        Configuration testedCompileConfiguration = variant.hasProperty("testedVariant") ?
+            variant.testedVariant.compileConfiguration : null
+        def artifactType = Attribute.of("artifactType", String)
+        def extractIncludeProtosTask =
+            maybeAddExtractIncludeProtosTask(
+                    variant.name,
+                    compileConfiguration.incoming.artifactView{ attributes{ it.attribute(artifactType, "jar") }}.files,
+                    testedCompileConfiguration == null ?
```
Comment (rschiu) [inline src/main/groovy/com/google/protobuf/gradle/ProtobufPlugin.groovy]: done

**Comment by rschiu** [comment]
@zhangkun83 Sorry, I didn't realized you did the review last week.  I made the changes.  PTAL.
The latest plugin requires newer buildtools.  The updated test uses compileSdk 25 and build tools 25.0.2, but I do not know how to update the SDK for the tests.  Can you help me with that?

**Comment by zhangkun83** [comment]
@rschiu I probably know less about Android SDK than you are. You can select from the SDK UI and install the packages, can't you?
Btw, when you push new changes to the PR, just do a normal commit, do not squash/rebase. This way the review history can be preserved.

**Comment by LouisCAD** [comment]
@rschiu You can now use SDK 26 and build tools `26.0.0`

File: src/main/groovy/com/google/protobuf/gradle/ProtobufPlugin.groovy [hunk 5]
```diff
@@ -241,9 +244,26 @@ class ProtobufPlugin implements Plugin<Project> {
       sourceSetNames.each { sourceSetName ->
         def extractProtosTask = maybeAddExtractProtosTask(sourceSetName)
         generateProtoTask.dependsOn(extractProtosTask)
+      }
 
-        def extractIncludeProtosTask = maybeAddExtractIncludeProtosTask(sourceSetName)
+      // Extract proto file from the compile classpath configuration exposed in Android Gradle plugin 2.5.  Revert to
+      // original behavior if the configuration is not found.
```
Comment (zhangkun83) [inline src/main/groovy/com/google/protobuf/gradle/ProtobufPlugin.groovy]: Just say the if-branch is for Android plugin 2.5+, and the else-branch is for below 2.5.
There is no "original" behavior outside the context of this PR.

**Comment by zhangkun83** [comment]
@rschiu I tested your PR with Android tools installed. It seems `compileArmFreeappDebugJavaWithJavac`  doesn't have the generated files in its input:
```
org.gradle.testkit.runner.UnexpectedBuildFailure: Unexpected build execution failure in /usr/local/google/home/zhangkun/github/protobuf-gradle-plugin-upstream/build/tests/testProjectAndroid with arguments [-DANDROID_PLUGIN_VERSION=3.0.0-alpha4, testProjectAndroid:build]

Output:
The com.google.protobuf plugin was already applied to the project: :testProjectAndroid and will not be applied again after plugin: android
Configuration 'compile' in project ':testProjectAndroid' is deprecated. Use 'implementation' instead.
:testProjectAndroid:preBuild UP-TO-DATE
:testProjectAndroid:preArmFreeappDebugBuild
Download https://plugins.gradle.org/m2/io/grpc/grpc-core/maven-metadata.xml
:testProjectAndroid:compileArmFreeappDebugAidl
:testProjectAndroid:compileArmFreeappDebugRenderscript
:testProjectAndroid:checkArmFreeappDebugManifest
:testProjectAndroid:generateArmFreeappDebugBuildConfig
:testProjectAndroid:extractArmFreeappDebugProto
:testProjectAndroid:extractArmProto
:testProjectAndroid:extractDebugProto
:testProjectAndroid:extractFreeappProto
:testProjectLite:extractIncludeProto
:testProjectLite:extractProto
:testProjectLite:generateProto
:testProjectLite:compileJavawarning: [options] bootstrap class path not set in conjunction with -source 1.7
Note: /usr/local/google/home/zhangkun/github/protobuf-gradle-plugin-upstream/build/tests/testProjectAndroid/testProjectLite/build/generated/source/proto/main/javalite/io/grpc/testing/Messages.java uses unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.
1 warning

:testProjectLite:processResources
:testProjectLite:classes
:testProjectLite:jar
:testProjectAndroid:extractIncludeArmFreeappDebugProto
:testProjectAndroid:extractProto
:testProjectAndroid:generateArmFreeappDebugProto
:testProjectAndroid:generateArmFreeappDebugResValues
:testProjectAndroid:generateArmFreeappDebugResources
:testProjectAndroid:mergeArmFreeappDebugResources
:testProjectAndroid:createArmFreeappDebugCompatibleScreenManifests
:testProjectAndroid:processArmFreeappDebugManifest
:testProjectAndroid:splitsDiscoveryTaskArmFreeappDebug
:testProjectAndroid:processArmFreeappDebugResources
:testProjectAndroid:generateArmFreeappDebugSources
:testProjectAndroid:javaPreCompileArmFreeappDebug
:testProjectAndroid:compileArmFreeappDebugJavaWithJavac/usr/local/google/home/zhangkun/github/protobuf-gradle-plugin-upstream/build/tests/testProjectAndroid/testProjectAndroid/build/generated/source/proto/armFreeappDebug/javalite/io/grpc/helloworldexample/Helloworld.java:104: error: package nano.Messages does not exist
    private nano.Messages.SimpleContext context_;
                         ^
/usr/local/google/home/zhangkun/github/protobuf-gradle-plugin-upstream/build/tests/testProjectAndroid/testProjectAndroid/build/generated/source/proto/armFreeappDebug/javalite/io/grpc/helloworldexample/Helloworld.java:122: error: package nano.Messages does not exist
    public nano.Messages.SimpleContext getContext() {
                        ^
/usr/local/google/home/zhangkun/github/protobuf-gradle-plugin-upstream/build/tests/testProjectAndroid/testProjectAndroid/build/generated/source/proto/armFreeappDebug/javalite/io/grpc/helloworldexample/Helloworld.java:132: error: package nano.Messages does not exist
    private void setContext(nano.Messages.SimpleContext value) {
                                         ^
/usr/local/google/home/zhangkun/github/protobuf-gradle-plugin-upstream/build/tests/testProjectAndroid/testProjectAndroid/build/generated/source/proto/armFreeappDebug/javalite/io/grpc/helloworldexample/Helloworld.java:147: error: package nano.Messages.SimpleContext does not exist
        nano.Messages.SimpleContext.Builder builderForValue) {
                                   ^
/usr/local/google/home/zhangkun/github/protobuf-gradle-plugin-upstream/build/tests/testProjectAndroid/testProjectAndroid/build/generated/source/proto/armFreeappDebug/javalite/io/grpc/helloworldexample/Helloworld.java:158: error: package nano.Messages does not exist
    private void mergeContext(nano.Messages.SimpleContext value) {
                                           ^
/usr/local/google/home/zhangkun/github/protobuf-gradle-plugin-upstream/build/tests/testProjectAndroid/testProjectAndroid/build/generated/source/proto/armFreeappDebug/javalite/io/grpc/helloworldexample/Helloworld.java:349: error: package nano.Messages does not exist
      public nano.Messages.SimpleContext getContext() {
                          ^
/usr/local/google/home/zhangkun/github/protobuf-gradle-plugin-upstream/build/tests/testProjectAndroid/testProjectAndroid/build/generated/source/proto/armFreeappDebug/javalite/io/grpc/helloworldexample/Helloworld.java:359: error: package nano.Messages does not exist
      public Builder setContext(nano.Messages.SimpleContext value) {
                                             ^
/usr/local/google/home/zhangkun/github/protobuf-gradle-plugin-upstream/build/tests/testProjectAndroid/testProjectAndroid/build/generated/source/proto/armFreeappDebug/javalite/io/grpc/helloworldexample/Helloworld.java:372: error: package nano.Messages.SimpleContext does not exist
          nano.Messages.SimpleContext.Builder builderForValue) {
                                     ^
/usr/local/google/home/zhangkun/github/protobuf-gradle-plugin-upstream/build/tests/testProjectAndroid/testProjectAndroid/build/generated/source/proto/armFreeappDebug/javalite/io/grpc/helloworldexample/Helloworld.java:384: error: package nano.Messages does not exist
      public Builder mergeContext(nano.Messages.SimpleContext value) {
                                               ^
/usr/local/google/home/zhangkun/github/protobuf-gradle-plugin-upstream/build/tests/testProjectAndroid/testProjectAndroid/build/generated/source/proto/armFreeappDebug/javalite/io/grpc/helloworldexample/Helloworld.java:40: error: package nano.Messages does not exist
    nano.Messages.SimpleContext getContext();
                 ^
/usr/local/google/home/zhangkun/github/protobuf-gradle-plugin-upstream/build/tests/testProjectAndroid/testProjectAndroid/build/generated/source/proto/armFreeappDebug/javalite/io/grpc/helloworldexample/Helloworld.java:123: error: package nano.Messages does not exist
      return context_ == null ? nano.Messages.SimpleContext.getDefaultInstance() : context_;
                                             ^
/usr/local/google/home/zhangkun/github/protobuf-gradle-plugin-upstream/build/tests/testProjectAndroid/testProjectAndroid/build/generated/source/proto/armFreeappDebug/javalite/io/grpc/helloworldexample/Helloworld.java:160: error: package nano.Messages does not exist
          context_ != nano.Messages.SimpleContext.getDefaultInstance()) {
                                   ^
/usr/local/google/home/zhangkun/github/protobuf-gradle-plugin-upstream/build/tests/testProjectAndroid/testProjectAndroid/build/generated/source/proto/armFreeappDebug/javalite/io/grpc/helloworldexample/Helloworld.java:162: error: package nano.Messages does not exist
          nano.Messages.SimpleContext.newBuilder(context_).mergeFrom(value).buildPartial();
                       ^
/usr/local/google/home/zhangkun/github/protobuf-gradle-plugin-upstream/build/tests/testProjectAndroid/testProjectAndroid/build/generated/source/proto/armFreeappDebug/javalite/io/grpc/helloworldexample/Helloworld.java:456: error: package nano.Messages.SimpleContext does not exist
                  nano.Messages.SimpleContext.Builder subBuilder = null;
                                             ^
/usr/local/google/home/zhangkun/github/protobuf-gradle-plugin-upstream/build/tests/testProjectAndroid/testProjectAndroid/build/generated/source/proto/armFreeappDebug/javalite/io/grpc/helloworldexample/Helloworld.java:460: error: package nano.Messages does not exist
                  context_ = input.readMessage(nano.Messages.SimpleContext.parser(), extensionRegistry);
                                                            ^
Note: /usr/local/google/home/zhangkun/github/protobuf-gradle-plugin-upstream/build/tests/testProjectAndroid/testProjectAndroid/src/main/java/io/grpc/helloworldexample/HelloworldActivity.java uses or overrides a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: Some input files use unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.
15 errors
 FAILED
```

**Comment by rschiu** [comment]
@zhangkun83 How did you run the test to get that build failure?  I have updated the tools to 26, but wasn't able to reproduce it.

**Comment by zhangkun83** [comment]
Did your run succeed instead?I just installed various SDK packages and ran `./gradlew clean test`.
I got failures first telling me some packages were missing. I installed them and was able to get the compile errors.

**Comment by rschiu** [comment]
Ya, I re-ran the tests and all runs succeeded.  I will dive deeper into it.

**Comment by LouisCAD** [comment]
@rschiu Could you provide an example of `before/after` gradle script for the workaround you describe on your first message of this PR, please?

**Comment by LouisCAD** [comment]
@rschiu What is blocking the progress of this PR and issue #129?
Anything I can do to help the Android Studio 3.0 compatible version come sooner?

**Comment by rschiu** [comment]
@zhangkun83 I have updated to the latest Android Gradle plugin.  Tests are passing on Travis, can you take another look?

**Comment by rschiu** [comment]
@LouisCAD I am not sure what you mean by a before/after gradle script for the workaround.  The change described in the first message in implemented in this PR.

**Comment by zhangkun83** [comment]
@rschiu Have you tried testing locally? I keep getting the same failure as mentioned above when testing it locally.

**Comment by rschiu** [comment]
Yes, the test pass locally as well.

**Comment by zhangkun83** [comment]
Hmm, it succeeded on my workstation after I `git clean`. Seems some left-over files that are not cleaned up by `gradle clean`.

File: src/test/groovy/com/google/plugins/ProtobufAndroidPluginTest.groovy [hunk 6]
```diff
@@ -15,10 +15,30 @@ class ProtobufAndroidPluginTest extends Specification {
...
+    repositories {
+        maven { url "https://plugins.gradle.org/m2/" }
+        if (androidPluginVersion.startsWith("3.")) {
+            google()
+        }
+    }
+    dependencies {
+        classpath "com.android.tools.build:gradle:$androidPluginVersion"
+    }
+}
```
Comment (zhangkun83) [inline src/test/groovy/com/google/plugins/ProtobufAndroidPluginTest.groovy]: This snippet is already in `testProjectAndroid/build.gradle`. Should it be removed?

**Comment by zhangkun83** [comment]
cc @zpencer

File: src/test/groovy/com/google/plugins/ProtobufAndroidPluginTest.groovy (hunk 6 above)
Comment (rschiu) [inline src/test/groovy/com/google/plugins/ProtobufAndroidPluginTest.groovy]: done
This is needed in the root project.  Remove the buildscript in testProjectAndroid/build.gradle

**Comment by rschiu** [comment]
Fixed an issue with the test buildscript.  PTAL.

File: src/test/groovy/com/google/plugins/ProtobufAndroidPluginTest.groovy [hunk 7]
```diff
@@ -15,10 +15,31 @@ class ProtobufAndroidPluginTest extends Specification {
     def mainProjectDir = ProtobufPluginTestHelper.prepareTestTempDir('testProjectAndroid')
     ProtobufPluginTestHelper.copyTestProjects(mainProjectDir, 'testProject', 'testProjectLite', 'testProjectAndroid')
 
+    // Add android plugin to the test root project so that Gradle an resolve
```
Comment (zpencer) [inline src/test/groovy/com/google/plugins/ProtobufAndroidPluginTest.groovy]: spelling: /an/can/

**Comment by zhangkun83** [comment]
@rschiu The test failed on Travis. It seems to be flaky. Please merge following to the `test` section of `build.gradle`, so that more details will be printed for the error.
```gradle
test {
    testLogging {
        exceptionFormat = 'full'
        showExceptions true
        showCauses true
        showStackTraces true
    }
}
```

**Comment by rschiu** [comment]
I have tried quite a few things, including running my own docker instance, but I am unable to reproduce the error locally.  Is anyone else able to do this on their local machine?  Any ideas?

**Comment by nmulcahey** [comment]
It runs locally for me, I suggest clearing the Travis CI cache and trying to run the build again.  $HOME/.gradle may also be too aggressive of a cache anyways.

https://docs.travis-ci.com/user/caching/#Clearing-Caches

**Comment by rschiu** [comment]
I assume you would need correct access to do that.  Could someone on the protobuf team try clearing the cache?

**Comment by zhangkun83** [comment]
I cleared the cache and re-ran the test. It's still broken.

**Comment by rschiu** [comment]
I was finally able to reproduce the problem with a docker instance limiting the memory.  So this seems to be a memory issue.  Is it possible to increase the memory size of the build server?

**Comment by zhangkun83** [comment]
Currently Travis is allocating [4GB of memory](https://docs.travis-ci.com/user/reference/precise/#Virtualization-environments). How much memory did you allocate in your docker instance? I wonder if you need to really increase the memory, or just tune up the java heap size. Anyway, it seems the way to have memory in Travis is to change `sudo: false` to `sudo: required` in `.travis.yml`

**Comment by rschiu** [comment]
That's kind of odd.  Is there anything else that affect the available memory?  I reproduced the problem when I limit the memory from the default of 2g to 1.5g.
At some point, I also tried to increase Gradle's heap size from 1g to 1.5g, but the daemon got killed because of that https://travis-ci.org/google/protobuf-gradle-plugin/builds/255465047

**Comment by rschiu** [comment]
Ok, I changed to sudo:required to get VM with larger memory and it's working now.  The tests ran successfully twice in a row, the previous failure was due to codenarc.  PTAL

**Comment by zhangkun83** [comment]
LGTM

**Pull request merged**